CDAP-13094 add gce dataproc provisioner

albertshau opened the pull request.

albertshau committed:
CDAP-13094 add gce dataproc provisioner

chtyim commented on cdap-runtime-ext-dataproc/pom.xml:
```diff
@@ -0,0 +1,113 @@
+    <dependency>
+      <groupId>com.google.cloud</groupId>
+      <artifactId>google-cloud-dataproc</artifactId>
+      <version>0.41.0-alpha</version>
```
Is that the only version can use? Using alpha seems a bit risky.

albertshau replied:
yeah they only have alpha versions

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java:
```diff
@@ -0,0 +1,167 @@
+        throw handleApiException((ApiException) cause);
+      }
+      throw Throwables.propagate(e);
+    }  catch (InterruptedException e) {
```
extra space

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java:
```diff
@@ -0,0 +1,167 @@
+      }
+      throw Throwables.propagate(e);
+    }  catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```
If you interrupt the current thread again, meaning the caller thread of this method would still get interrupted (e.g. on calling sleep inside RetryStrategy), is the what you wanted?

albertshau replied:
I wasn't quite sure what to do, but now that I think about it some more, I think the right thing is not to catch it and let the caller handle it.

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java:
```diff
@@ -0,0 +1,167 @@
+
+  @Nullable
+  public OperationSnapshot deleteCluster(String name) throws RetryableProvisionException {
+
```
extra empty line

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java:
```diff
@@ -0,0 +1,167 @@
+      Thread.currentThread().interrupt();
+      throw new RetryableProvisionException(e);
+    } catch (TimeoutException e) {
+      throw new RetryableProvisionException(e);
```
Is it safe to issue the `deleteClusterAsync` again on timeout of the `getInitialFuture.get` call? Why can't the `get` block indefinitely?

albertshau replied:
hm I suppose it's not correct. For the timeout, I was worried about environments where these public APIs are blocked from the network and these requests hang forever, but in those situations the request should not be retried.

albertshau replied:
that would be true of the getCluster() call too though (which doesn't use a future), so the caller must have a way to timeout. I'll just change it to a get.

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java:
```diff
@@ -0,0 +1,167 @@
+      }
+      throw Throwables.propagate(e);
+    }  catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
```
Same as above.

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java:
```diff
@@ -0,0 +1,167 @@
+    }
+  }
+
+  @Nullable
```
In Java 8, consider using `Optional<Cluster>` as return type instead of null.

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java:
```diff
@@ -0,0 +1,167 @@
+    this.client = client;
+  }
+
+  public OperationSnapshot createCluster(String name) throws RetryableProvisionException {
```
Some javadoc would be helpful (e.g. when it will throw what exception).

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcTool.java:
```diff
@@ -0,0 +1,118 @@
+/**
+ * Manual tool for testing out dataproc provisioning and deprovisioning.
+ */
+public class DataProcTool {
```
Not need now, but it would be nice if there is a SPI to have the tool integrated with CDAP cli. Probably open a JIRA for that.

albertshau replied:
opened CDAP-13244

albertshau committed:
cleanup around the DataProcClient

Using Optional instead of nulls, adding javadocs, and fixing
exception handling.

albertshau commented:
https://builds.cask.co/browse/CDAP-DUT6351

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcClient.java:
```diff
@@ -0,0 +1,180 @@
+        throw new RetryableProvisionException(e);
+      }
+      // otherwise, it's not a retryable failure
+      throw e;
```
Is `ApiException` a `RuntimeException`?

albertshau replied:
yeah

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcProvisioner.java:
```diff
@@ -0,0 +1,134 @@
+      // if it already exists, it means this is a retry. We can skip actually making the request
+      Optional<com.google.cloud.dataproc.v1.Cluster> existing = client.getCluster(clusterName);
+
+      // TODO: figure out how to get this if existing is not null
```
What is the TODO for?

albertshau replied:
this situation is if we're retrying this call and the cluster already exists, then we should use the create time for the original create request, not the current time.

chtyim commented on cdap-runtime-ext-dataproc/src/main/java/co/cask/cdap/runtime/spi/provisioner/dataproc/DataProcProvisioner.java:
```diff
@@ -0,0 +1,134 @@
+      long createTime = System.currentTimeMillis();
+
+      if (!existing.isPresent()) {
+        client.createCluster(clusterName);
```
So we don't need the result of `createCluster`?

albertshau replied:
Doesn't seem like anything that comes back is required for future operations.

albertshau committed:
look up cluster create time from an existing cluster

chtyim reviewed (approved):
LGTM
